### PR TITLE
feat: Support Gemma3-1b-it in Pytorch workflow

### PR DESCRIPTION
**Pull request by brb-nv**
## Description
This MR adds support for Gemma3-1b-it in Pytorch workflow. There's scope for KV Cache optimization using Cyclic Cache for Variable Sliding Window Attention that will be addressed in a future MR. Plugging in `max_attention_window_vec` in KVCacheManager config didn't suffice and yielded poor results.

```
$ python3 examples/pytorch/quickstart_advanced.py --model_dir /home/scratch.trt_llm_data/llm-models/gemma/gemma-3-1b-it/ --max_tokens 2048 --max_seq_len 2048 --prompt "Write me a really long Shakespeare play."
```

## Test Coverage

```
$ pytest tests/integration/defs/accuracy/test_llm_api_pytorch.py::TestGemma3_1BInstruct::test_auto_dtype -s -v
```

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--disable-fail-fast --skip-test --stage-list "A10-1, xxx" --gpu-type "A30, H100_PCIe" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-[Post-Merge]-1, xxx"]`

Launch build/test pipelines. All previously running jobs will be killed.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests. Will also run L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-[Post-Merge]-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-[Post-Merge]-1, xxx".

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

File: tensorrt_llm/_torch/models/__init__.py [hunk 1]
```diff
@@ -27,6 +28,7 @@
     "BertForSequenceClassification",
     "CLIPVisionModel",
     "DeepseekV3ForCausalLM",
+    "Gemma3ForCausalLM",
```
Comment (amukkara) [inline tensorrt_llm/_torch/models/__init__.py:31]: update model support table in `examples/pytorch/README.md`

File: tensorrt_llm/_torch/models/modeling_utils.py [hunk 2]
```diff
@@ -693,4 +693,8 @@ def filter_weights(prefix, weights: Dict):
                 else:
                     for n, p in module._parameters.items():
                         if p is not None:
-                            p.data.copy_(module_weights[n][:])
+                            # Hack for Gemma3 RMSNorm.
```
Comment (amukkara) [inline tensorrt_llm/_torch/models/modeling_utils.py:696]: please add a detailed comment why this hack is needed.

File: tensorrt_llm/_torch/models/modeling_gemma3.py [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+        attn_metadata: AttentionMetadata,
+        input_ids: torch.LongTensor = None,
+        position_ids: Optional[torch.LongTensor] = None,
+        inputs_embeds: Optional[torch.FloatTensor] = None,
+        pipeline_interface: Optional[PipelineInterface] = None,
+        return_context_logits: bool = False,
+        **kwargs,
+    ) -> torch.Tensor:
+
+        if self._supports_pp and self.pp_size > 1:
```
Comment (amukkara) [inline tensorrt_llm/_torch/models/modeling_gemma3.py:255]: FYI https://github.com/NVIDIA/TensorRT-LLM/pull/4034 is refactoring PP implementation. you will need to modify this if that PR is merged first.

File: tensorrt_llm/_torch/models/__init__.py [hunk 4]
```diff
@@ -4,6 +4,7 @@
 from .modeling_bert import BertForSequenceClassification
 from .modeling_clip import CLIPVisionModel
 from .modeling_deepseekv3 import DeepseekV3ForCausalLM
+from .modeling_gemma3 import Gemma3ForCausalLM
```
Comment (amukkara) [inline tensorrt_llm/_torch/models/__init__.py:7]: add test cases.

File: tensorrt_llm/_torch/models/__init__.py (hunk 1 above)
Comment (brb-nv) [reply tensorrt_llm/_torch/models/__init__.py:31]: Done: https://github.com/NVIDIA/TensorRT-LLM/pull/3999/commits/e6021433c48d85e0c567b670399f5d5a0af2e29d

File: tensorrt_llm/_torch/models/__init__.py (hunk 4 above)
Comment (brb-nv) [reply tensorrt_llm/_torch/models/__init__.py:7]: Done. https://github.com/NVIDIA/TensorRT-LLM/pull/3999/commits/c895915cbd3dc9c80dc602d3cce2a8561d30ef39

File: tensorrt_llm/_torch/models/modeling_gemma3.py (hunk 3 above)
Comment (brb-nv) [reply tensorrt_llm/_torch/models/modeling_gemma3.py:284]: Noted. Thank you for letting me know, Anurag.

File: tensorrt_llm/_torch/models/modeling_utils.py (hunk 2 above)
Comment (brb-nv) [reply tensorrt_llm/_torch/models/modeling_utils.py]: Removed the hack from modeling_utils and wrote custom weight loading logic for gemma3 where RMSNorm weight loading is duly handled.
https://github.com/NVIDIA/TensorRT-LLM/pull/3999/commits/dfa8b24f616c0818c2680feaf453262f396f6ec1

File: tests/integration/defs/accuracy/test_llm_api_pytorch.py [hunk 5]
```diff
@@ -252,6 +252,16 @@ def test_auto_dtype(self):
             task.evaluate(llm)
 
 
+class TestGemma3(LlmapiAccuracyTestHarness):
```
Comment (amukkara) [inline tests/integration/defs/accuracy/test_llm_api_pytorch.py:255]: since more gemma-3 models will be added later, do you want to add model size tag (1b) in test name?

**Review by amukkara — APPROVED** [review]

File: tests/integration/defs/accuracy/test_llm_api_pytorch.py (hunk 5 above)
Comment (brb-nv) [reply tests/integration/defs/accuracy/test_llm_api_pytorch.py]: Sounds good. Updated.
